### ISSUE TITLE
Permissions Policy for WebGPU and for unmaskHints

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#3483]
In order for WebGPU, or `unmaskHints`, to be gated by Permissions Policy (aka Feature Policy), I think we need to specify it. There's currently an inline issue in [Abuse of capabilities](https://gpuweb.github.io/gpuweb/#security-abuse-of-capabilities) about using Permissions Policy as a mitigation, but I think if we want to do that, it needs to be standardized and specified.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Feature_Policy/Using_Feature_Policy

As described here, each feature has a default allowlist. As mentioned in that inline issue, our default allowlist might be `'self'`:
> The feature will be allowed in this document, and in all nested browsing contexts (iframes) in the same origin.

- [ ] Come up with a proposal
- [ ] Iron it out with policy experts
- [ ] Add to our spec at the affected points, plus somewhere under "malicious use considerations"
- [ ] Add to https://github.com/w3c/webappsec-permissions-policy/blob/main/features.md

**Comment by kainino0x** [comment]
As an example, the geolocation spec defines its token string and default allowlist here:

https://www.w3.org/TR/geolocation/#permissions-policy

**Comment by kainino0x** [comment]
This is part of #3101, I forgot.

**Comment by kainino0x** [comment]
I've thought of several levels of permission we could add. I think we could have as many as all of these - so what we choose depends on how many levels we want to provide.

- Permission to use WebGPU at all
- Permission to use WebGPU with hardware acceleration (fingerprintable with rendering results unless noise is injected)
- Permission to use WebGPU with more than the base capabilities
- Permission to use unmaskHints in requestAdapterInfo

requestAdapterInfo could likely create a permission prompt and therefore is probably extra important to provide a permission policy for. requestAdapter can _theoretically_ create a permission prompt, but it's not intended to, so I think choices for the others are more subjective.

The design constraints:

- Post-v1 we can't break existing pages so we can't put things behind permissions that were not before.
- We _can_ add permissions that expose things that weren't exposed before. For example, we could have a "webgpu-restricted" permission in v1 and then a "webgpu-unrestricted" permission post-v1.
    - It's also possible to have a "webgpu-unrestricted" permission in v1, and add a "webgpu-restricted" post-v1 without breaking anything. But it's also not ideal, because old code will grant more permissions than they need to. (I think this describes @litherum's concern from the meeting).

**Comment by Kangz** [comment]
In the last meeting the group agreed that forward progress on this issue would be through specific PRs (non-normative text can be added to WebGPU before normative text is added to the HTML spec)

**Comment by saccharineboi** [comment]
> I've thought of several levels of permission we could add.

Each n-th permission adds extra bit of info which could be used for fingerprinting. If there are 6 levels, that's 6 bits of identifying info. This could be used to cluster users for targeted attacks.

This looks like a classical case of "let us protect our users from malicious software" mindset that actually ends up being malicious itself.

If this gets passed there should at least be a paragraph about it in "2.2. Privacy Considerations".

**Comment by kainino0x** [comment]
That's useful to know. We are not currently considering adding all of these levels, just trying to figure out which 1-2 to choose from that list, but perhaps this info pushes us toward 1. In any case, we intend to come up with a proposal, but whatever we end up with will be designed with help from policy experts. Which reminds me a few steps so I added a checklist to this issue.

FTR:
- [Minutes from 2022-10-19](https://github.com/gpuweb/gpuweb/wiki/Minutes-2022-10-19#permissions-policy-for-webgpu-and-for-unmaskhints-3483)
- [Minutes from 2022-10-26](https://github.com/gpuweb/gpuweb/wiki/Minutes-2022-10-26#permissions-policy-for-webgpu-and-for-unmaskhints-3483)

**Comment by Kangz** [comment]
@saccharineboi can you detail, or do you have pointer to an explanation of why these would add bits of fingerprinting? It seems that the parent website would pass them into the [`"allow" iframe attribute`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Feature_Policy/Using_Feature_Policy#the_iframe_allow_attribute) so it's not exactly like permissions granted by the user that are bits of fingerprinting vector. Here the allow-list would most likely depend on the embedder website that can already be queried in some ways through `window.parent` no?

**Comment by kainino0x** [comment]
Here's a concrete proposal:

**Without the permission granted:**
- There can never be a prompt for `requestAdapterInfo()`.
- Browser **may** impose additional WebGPU restrictions compared to what the host page would get, e.g.:
    - Only the core API (or a subset of limits/features which are so widespread that they're not generally useful for fingerprinting)
    - A software implementation

**With the permission granted:**
- The iframe can cause the host page to trigger a prompt for `requestAdapterInfo()`.
- Browser **should not** impose additional WebGPU restrictions.

---

This will need some scrutiny. There's one thing I'm not sure about:
- Aside from the possibility of a requestAdapterInfo() prompt, this policy is just a hint. I think the possibility of that prompt may justify a policy (even though no one expects to implement a prompt for it yet). But does it make sense to roll a bunch of hints up with it? I think it does.

And we need to figure out _who_ can provide that scrutiny. Is it the PING? I'll find out.

**Comment by domenic** [comment]
> Each n-th permission adds extra bit of info which could be used for fingerprinting. If there are 6 levels, that's 6 bits of identifying info. This could be used to cluster users for targeted attacks.

This isn't quite correct. Permissions policy isn't related to user-facing and user-toggleable permissions. It's rather a way for pages to control how what permissions they already have, are delegated to iframes.

In particular, I don't think anyone is proposing adding a user-facing permission (and thus fingerprinting bit) for WebGPU. I think this issue is just about whether to add a permissions policy to allow top-level pages to control the WebGPU abilities of cross-origin iframes, or whether to use the web platform's default approach where cross-origin iframes can do everything the same as the top-level page.

**Comment by kainino0x** [comment]
Ah, I understand the concern better now, thank you! Indeed, we're not proposing the addition of any user-facing permissions. There is one mentioned - for requestAdapterInfo - but that (\*very\* hypothetical) permission prompt would not be just for "page wants to know your GPU info", but a very general "page might be able to identify your device". We're just leaving the possibility open for a prompt, or, to speculate wildly, something like a browser-global setting for "more private" vs "less private", both of which would be fingerprinting bits by Domenic's explanation.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/Minutes-2023-01-04#permissions-policy-for-webgpu-and-for-unmaskhints-3483">GPU Web 2023-01-04/05 APAC-timed</a></summary>

* KN(pretyped): have some feedback from our spec mentors. Also if we were to do user activation of some kind (see #3101) it could play in here
* KN: gist of feedback: probably shouldn't spec a permission policy for something with non-normative effects. One I proposed says, if you don't have the permission, the browser may impose add'l restrictions in the iframe. Domenic said, if we aren't confident in spec'ing specific behavior, probably an interop risk to do that. Fair. People already have to deal with different device capabilities, so I think risk's reduced, but fair point.
* KN: he did say, allowing top-level origin to restrict cross-origin iframes from doing things makes sense. Good to have the ability. Not sure how to limit it more tightly.
* MM: is the scope of this issue just about interactions between frames & sub-frames? Or also including interactions with Permissions API or others?
* KN: Don't remember what Permissions API looks exactly like. Allows a page to delegate caps to x-origin iframes it embeds. Also exists at HTTP level. We'd have a default to let pages use WebGPU. You can use headers to disable a particular page's WebGPU access. Don't know how this interacts with Permissions API. (iframe can check what's delegated to it.)
* MM: Also gives code location for prompts to appear to user. Scary - already have such a place.
* KN: if you requestAdapter multiple times, we'll prompt you on the first one. If you do permissions.request(), that'd count as the first one. MDN also says that's experimental. permissions.query() is the only one you can use. Others are deprecated?
* MM: it's more interesting to us how it interacts with the permissions API than iframes.
* KG: permissions policy's only for frame-to-frame.
* KN: essentially yes. TR version of permissions API doesn't have anything except for query().
* KN: also: Domenic said, if there were to be a prompt, you'd want a permission so host page can delegate to iframe the ability to show a prompt - important. But since we aren't planning for a prompt, not sure how we'd expose it.
* KN: prompt I imagine - identify your machine. How'd you write it so a user can understand it? Wouldn't be WebGPU specific.
* KG: have had 2 ideas - I want debug info for this machine because something went wrong. Or, something to give pages/origins unchecked access, like installed apps have root access. Not something we're pursuing though. Also, I don't think most apps that want the renderer info need to know specifically. Giving them coarse-grained info is good enough. I don't think there's value between the coarse- and fine-grained info
* KN: for requestAdapterInfo I agree. But, exposing fewer features, a smaller adapter, or no adapter to iframes - seems generally desirable. Prevent embeds from doing "interesting" things.
* MM: do implementers think iframes should get WebGPU by default?
* KG: yes.
* KR: nods.
* KN: yes.
* MM: I meant third-party iframes.
* KN: I think so but no strong reason.
* KG: reason in my mind - is this part of the web? Or, someday, will we have third-party iframes be lesser citizens of the web API? If no current plan there, not sure why WebGPU would have that plan. Everyone doing graphics might spin up WebGL instead of WebGPU.
* KR: I think not supporting webgpu by default would damage ecosystem for e.g. sketchfab on twitter etc. I think we should have webgpu avail by default, shouldn’t segregate it off. It's primarily a visualization API.
* KN: fair point. Didn't think about iframes already existing and not delegating this permission.
* KG: any stakeholders left standing that want us to do a permissions policy for WebGPU V1? Fine if not.
* KN: not 100% sure personally. Despite concerns with my proposal I still like it - like having the option even if it's a slight interop risk. Better than the backward compat risk if we want to add it later.
* KG: then the question is, are we adding a skeleton that doesn't do anything right now?
* KN: wouldn't want to do that - same as not doing anything. People would have to implement something. Maybe you only get a core adapter.
* KG: doesn't really address abusive capabilities spec. Mainly talks about compute shaders.
* KN: think I'm happy to not do a permissions policy. Will follow up with folks I was talking with.
* KR: Things that we’ve tried to lockdown in webgl are like, access dgpu from 3rd party iframe. We’ve changed policies on devs, maybe didn’t have a big impact because it’s just perf/power. [version numbers in webgl?] Adding a Perm Policy is probably worth it to allow people to opt-in to e.g. more-power-use if they want.
* KN: in Chrome we have had to add mitigations to existing APIs. Like, we detect ad iframes, and put limitations on it. Can do this heuristically. Has been proven out with past APIs. Not sure how critical it is to do better than previous APIs.
* KG: concerned about taking away tools from people if they're worried about their use in an iframe. A lot of uses of WebGL that aren't obviously WebGL. This app doesn't work anymore, why? Give us your preferences list, they disabled WebGL. Do you want people to develop a non-WebGPU version because it might be forbidden in iframes? If we're concerned about e.g. compute shaders we could potentially split that off, but don't think it solves the problem that much.
* KG: power usage - containing doc maybe knows how much power inner documents/iframes should be able to use. Ad networks shouldn't be able to consume 100% GPU, etc.
* GT: Kai brought up idea, main page can only use CPU-emulated version of WebGPU. That's more power, not less. Not a win for the main page. Pointing that out.
* KG: I'm thinking iGPU vs. dGPU mainly.
* KN: reason for CPU was for privacy, fingerprinting mitigations. Happy to say fingerprinting mitigations shouldn't be a goal of the permission policy.
* KG: permission policy for power use? Discuss? Goal for this issue isn't to have something to have a permissions policy for, but address the line item of abusive capabilities. If we don't think we need to do that, we should remove it and instead file an issue for e.g. power usage.
* KN: agree with closing this based on that outcome. Possibly opening a new issue.
* KG: think it's worth writing down the resolution.
* MM: to confirm - there will be no tokens that a parent frame can pass to a child frame to delegate WebGPU specific permissions.
* KN: we won't have one which lets you delegate "fingerprinting capability". Might consider one for power use.
* MM: OK. If someone comes later with new arguments, this issue could be reopened? Not making a philosophical decision, but a "for-now" decision?
* KN: yes. Planning to get more feedback before settling this forever.
* KG: this is however the last call for V1 on this topic.

</details>

**Comment by kainino0x** [comment]
To crystallize/broaden Ken's from the last CG discussion:

We discussed this for V1 because adding any restrictions would break iframes written after WebGPU 1.0, but before a new restriction was added. But what we didn't consider was that iframes written before WebGPU 1.0 would also be affected. We don't want to create a situation where engines need to maintain WebGL backends _just_ because they need to run in embeds - imagine someone embedded a Sketchfab in their website in 2020, and then in 2025 Sketchfab decided they wanted to drop their WebGL backend. Thus in a default iframe:
- WebGPU would need to be available
- WebGPU would need to be fast (not software)
- WebGPU would need to have a reasonable baseline of functionality, ideally at least as high as WebGL in the same iframe

We can break this if we want, but it's not looking like a good tradeoff.

**Comment by kainino0x** [comment]
There's an additional problem we've discussed in the past that doesn't seem to be documented here - trackers and ad networks are injected by `<script>`, which controls the attributes of the iframe it generates (if any), so having a permission policy doesn't do anything to protect the user's privacy.

Between this and the above problems, in the meeting we decided not to pursue this.
